Add support for creating self-decrypting binaries

will-v-pi opened the pull request:
This adds extra functionality to the `pico_encrypt_binary` function to allow creating self-decrypting binaries, including specifying the OTP page to use for the AES key.

```cmake
pico_encrypt_binary(hello_encrypted
    ${CMAKE_CURRENT_LIST_DIR}/privateaes.bin
    ${CMAKE_CURRENT_LIST_DIR}/ivsalt.bin
    EMBED
    OTP_KEY_PAGE 29)
```

The main non-backwards-compatible change is the addition of a new IV salt bin file which is required and must be passed as the second argument. This will break any uses of `pico_encrypt_binary`, and has been added for security purposes, as we now salt the public IV with a salt stored in OTP.

The other non-backwards-compatible change it that if you previously called:
```cmake
pico_encrypt_binary(my_target my_aesfile.bin my_sigfile.pem)
```
you now need to call
```cmake
pico_encrypt_binary(my_target my_aesfile.bin my_iv_salt.bin SIGFILE my_sigfile.pem)
```
due to the new argument parsing. I think that this is fine, because the only time you'd pass a separated `SIGFILE` to `pico_encrypt_binary` is when you're using a different signing key for the binary vs the encrypted blob, which is not a common use case.

This PR requires use of the picotool encrypted-shares branch (https://github.com/raspberrypi/picotool/pull/207), so should be merged after that.

will-v-pi committed:
Add pico_create_decrypting_binary function

For now, this function embeds the decrypting bootloader, but probably better to integrate (or replace) existing pico_encrypt_binary function

will-v-pi committed:
Add EMBED and OTP_KEY_PAGE arguments to pico_encrypt_binary function

will-v-pi committed:
Support passing otp_file to picotool encrypt

will-v-pi committed:
Add IV salt to pico_encrypt_binary

will-v-pi committed:
Add MBEDTLS option to use mbedtls decryption stage

will-v-pi committed:
Improve docs for pico_encrypt_binary

Add newlines for readability, and explain why MbedTLS version is insecure

lurch commented on tools/CMakeLists.txt:
```diff
@@ -380,19 +404,58 @@ function(pico_embed_pt_in_binary TARGET PTFILE)
+# Optionally, use EMBED to embed a decryption stage into the encrypted binary.
+# This sets PICOTOOL_EMBED_DECRYPTION to TRUE.
+#
+# Optionally, use MBEDTLS to to use the MbedTLS based decryption stage - this
+# is faster, but offers no security against power or timing sniffing attacks,
```
Should this caveat also be mentioned somewhere in the picotool PR related to this?

will-v-pi replied:
It'll be documented more thoroughly in the C SDK book, but I'll add a quick note to the help text for the option

lurch commented on tools/CMakeLists.txt:
```diff
@@ -380,19 +404,58 @@ function(pico_embed_pt_in_binary TARGET PTFILE)
-# file containing 32 bytes of a random key), and sign the encrypted binary.
-# This sets PICOTOOL_AESFILE to AESFILE, and PICOTOOL_ENC_SIGFILE to SIGFILE
-# if present, else PICOTOOL_SIGFILE.
-function(pico_encrypt_binary TARGET AESFILE)
+# file containing 128 bytes of a random key), and sign the encrypted binary.
```
I guess this is now a 128-byte file (key share) OR a 32 byte file (key) OR a 64-character hex string (key) ?

lurch commented on tools/CMakeLists.txt:
```diff
@@ -380,19 +404,58 @@ function(pico_embed_pt_in_binary TARGET PTFILE)
-# if present, else PICOTOOL_SIGFILE.
-function(pico_encrypt_binary TARGET AESFILE)
+# file containing 128 bytes of a random key), and sign the encrypted binary.
+# Salts the public IV with the provided IVFILE (should be a binary file
+# containing 16 bytes of a random IV), to give the IV used by the encryption.
```
And similarly I guess this is now a 16-byte file OR a 32-char hex string?

will-v-pi replied:
Yeah, I think I'll say that the SDK only supports passing files - you really shouldn't be putting a secret key in plain text in your CMakeLists.txt file

will-v-pi committed:
Key can be 128 byte key share or 32 byte key

lurch commented on tools/CMakeLists.txt:
```diff
@@ -380,19 +404,59 @@ function(pico_embed_pt_in_binary TARGET PTFILE)
 endfunction()
 
-# pico_encrypt_binary(TARGET AESFILE [SIGFILE])
+# pico_encrypt_binary(TARGET AESFILE IVFILE [SIGFILE <file>] [EMBED] [MBEDTLS] [OTP_KEY_PAGE <page>])
```
Is it worth preemptively adding your `\param\` stuff from #2422 here too, or is it too early for that?

will-v-pi replied:
I'll wait till that's merged, then rebase this and fix the issues that presumably pop up

will-v-pi replied:
Merged and rebased

lurch commented on tools/CMakeLists.txt:
```diff
@@ -420,25 +444,70 @@ function(pico_embed_pt_in_binary TARGET PTFILE)
+# This sets the target properties PICOTOOL_AESFILE to AESFILE, PICOTOOL_IVFILE to IVFILE, and
+# PICOTOOL_ENC_SIGFILE to SIGFILE if specified, else PICOTOOL_SIGFILE.
+#
+# Optionally, use EMBED to embed a decryption stage into the encrypted binary.
```
Another stupid question: when/why would you _not_ want to embed a decryption stage into the encrypted binary?

lurch commented on tools/CMakeLists.txt:
```diff
@@ -420,25 +444,70 @@ function(pico_embed_pt_in_binary TARGET PTFILE)
+# and takes up more code size.
+# This sets the target property PICOTOOL_USE_MBEDTLS_DECRYPTION to TRUE.
+#
+# Optionally, use OTP_KEY_PAGE to specify the OTP page storing the AES key.
```
Does this have some default or implicit value, if not explicitly set?

lurch replied:
Is this same OTP page (or some fixed offset from it) also used for storing the IV salt, or _might_ it make sense to have an `OTP_SALT_PAGE` parameter?

will-v-pi committed:
Always package encrypted binaries with embedded decryption stages

The embedded decryption stage only works if the binary is written to Flash and then loaded to SRAM by the bootrom

kilograham committed:
Merge branch 'develop' into self-decrypting-only
